### DISCUSSION TITLE
Esp32 s3 box Support?

### DISCUSSION THREAD

**Discussion by jouellnyc** [micropython/micropython#10391] [Hardware & Peripherals] [answered]
Howdy. And happy new year.

Wondering if any of the esp32 s3 boxes will receive support from MicroPython?

https://www.espressif.com/en/news/ESP32-S3-BOX_video

It seems like Circuit Python supports it so maybe that helps?

https://circuitpython.org/board/espressif_esp32s3_box_lite/

The boxes look like they'd be great for those looking to avoid soldering but want a nice presentable device. 

Was planning on buy and test for a project, so just trying to see what the future might look like for them.

I'll likely try one using circuit python soon. If you think it's helpful I could share my experience/details back here...

Thanks

**Comment by mattytrentini** [comment, marked as answer]
You can use the [GENERIC_S3_SPIRAM_OCT](https://micropython.org/download/GENERIC_S3_SPIRAM_OCT/) board; it'll run fine on the ESP32-S3-Box. It's equivalent to the CircuitPython firmware.

**Reply by jouellnyc** [reply]
Oh. Excellent to hear. Thanks  @mattytrentini !

**Comment by jouellnyc** [comment]
Howdy! I'm wondering if there is a suggested video driver for the ESP32-S3-Box? I have not quite gotten a reliable way to write to the ESP32-S3-Box screen, having tried a few repos.

**Reply by russhughes** [reply]
Have you tried [This](https://github.com/russhughes/st7789_mpy/blob/master/firmware/ESP32_BOX_LITE/firmware.bin) from https://github.com/russhughes/st7789_mpy ? Use the this [tft_config.py]( https://github.com/russhughes/st7789_mpy/blob/master/examples/configs/esp32_box_lite/tft_config.py) for the ESP32 Box and Box Lite's ili9342c display.

**Reply by jouellnyc** [reply]
Yes, I got it to flash once, but having challenges now. Post good Thonny/esptool flash and it reboot, 'Invalid header' shows repeatedly. But also I have a specific [XGLCD font ](https://www.mikroe.com/glcd-font-creator) need that was created from Windows. That font seems to crash the esp32 using st7789. I have been hesitant to raise an issue as I'm not sure the problem is which the st7789 repo per se. 

FWIW, that font does work w/[this](https://github.com/rdagger/micropython-ili9341) repo but up [until a day or so ago] https://github.com/rdagger/micropython-ili9341/issues/13) it wrote 'backwards and downwards' to the screen. (but original intent was not for the S3 Box). 

Project progression was from breadboard/OLED/ssd1306, to an ili93414/3D-printed housing to the ESP32-S3-Box/Lite. (Costs/overhead for a professional package/assembly is prohibitive given my lack of CAD/CAM knowledge). Once switched to *Boxes I have been pursuing "whichever worked best" with the intent of using the XGLCD font. Thanks so much for the reply. I'm happy to test/provide any feedback.

**Reply by russhughes** [reply]
The XGLCD font won't work as it uses a different encoding style.  A tool in the utils directory named `font2bitmap.py` converts TrueType fonts into python modules that will work with the firmware.  

Repeating 'invalid header' often indicates a bad flash or the need to erase the flash before uploading the firmware.bin. ESP32S3 firmware.bin files should be flashed starting at 0x0 and not at 0x1000 like you would for a Vanilla ESP32. The last time I used the Thonny firmware flash tool, it tried to flash starting at 0x1000. 

You should be able to modify the [ROTATE](https://github.com/rdagger/micropython-ili9341/blob/d080d5bac95c0da972b26e3599f56bb0311d9ebd/ili9341.py#L85) list in rdagger'a driver to get the orientations you need. Try the values I used in my firmware, which are:

```
    ROTATE = {
        0: 0xc8,
        90: 0xa8,
        180: 0x08,
        270: 0x68
    }
```

Russ

**Reply by jouellnyc** [reply]
The XGLCD font won't work as it uses a different encoding style. A tool in the utils directory named font2bitmap.py converts TrueType fonts into python modules that will work with the firmware.
>>Excellent to know. Thank you very much for the insight and the utility. Will check it out.

Repeating 'invalid header' often indicates a bad flash or the need to erase the flash before uploading the firmware.bin. ESP32S3 firmware.bin files should be flashed starting at 0x0 and not at 0x1000 like you would for a Vanilla ESP32. The last time I used the Thonny firmware flash tool, it tried to flash starting at 0x1000.

>> I share your thoughts in that.  I think it was because I was using the '-z' flag while flashing. It worked just now w/o using that flag.

You should be able to modify the [ROTATE](https://github.com/rdagger/micropython-ili9341/blob/d080d5bac95c0da972b26e3599f56bb0311d9ebd/ili9341.py#L85) list in rdagger'a driver to get the orientations you need. Try the values I used in my firmware, which are:

    ROTATE = {
        0: 0xc8,
        90: 0xa8,
        180: 0x08,
        270: 0x68
    }

>>Indeed! I went with:

ROTATE = {
        0: 0xC8,
        90: 0xA8,
        180: 0x48,
        270: 0x28
    }

>> With "0xC8" making it look 'normal'. 

>>Thanks again for the replies. This is all very encouraging.

**Reply by jouellnyc** [reply]
The firmware and config do work perfectly to write to the esp box screen, and flashing as suggested:
https://github.com/orgs/micropython/discussions/10391#discussioncomment-5203911
(This is super/excellent on it's own.)

Interestingly (to me) after I do the init() call, I am able to use the ili9341 display driver code I had before  and then use the XGLCD font as I was doing previously. I partly stumbled over this fact, but sharing here. 

Just looking at the code in the repo suggests quite a lot of work with the custom init etc, so I am guessing once the backlight is turned on (which the ili9341 display driver does not appear to do - mainly b/c it was not designed for the s3 box in mind), the ili9341 display driver is now able to display to the screen.

Again, many thanks for this https://github.com/orgs/micropython/discussions/10391#discussioncomment-5203911!
